### PR TITLE
Add a mutex and used ports list to the tests random port generator to avoid port collisions

### LINKED ISSUE #15071: `TestReadyAndHealthy` CI test flaky
### What did you do?

-

### What did you expect to see?

_No response_

### What did you see instead? Under which circumstances?

The `TestReadyAndHealthy` test failed in CI, seems flaky:

```
--- FAIL: TestReadyAndHealthy (0.02s)
panic: Unable to start web listeners: listen tcp :38639: bind: address already in use [recovered]
	panic: Unable to start web listeners: listen tcp :38639: bind: address already in use

goroutine 133 [running]:
testing.tRunner.func1.2({0x12cf780, 0xc000620b90})
	/usr/local/go/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1635 +0x35e
panic({0x12cf780?, 0xc000620b90?})
	/usr/local/go/src/runtime/panic.go:785 +0x132
github.com/prometheus/prometheus/web.TestReadyAndHealthy(0xc000032b60)
	/__w/prometheus/prometheus/web/web_test.go:106 +0x108b
testing.tRunner(0xc000032b60, 0x159b7b0)
	/usr/local/go/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1743 +0x390
FAIL	github.com/prometheus/prometheus/web	10.078s
```

### System information

_No response_

### Prometheus version

_No response_

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by jadolg**
Fixes #15071

### CODE REVIEW

File: util/testutil/port.go [hunk 1]
```diff
@@ -15,21 +15,56 @@ package testutil
...
+
+	for portWasUsed(port) {
+		port, err = getPort()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	usedPorts = append(usedPorts, port)
+	mu.Unlock()
```
Comment (juliusv) [inline util/testutil/port.go:45]: Let's do

```go
mu.Lock()
defer mu.Unlock()
```

at the top, that way we have to think less about unlocking correctness.

**Review by juliusv — COMMENTED** [review]
Thanks, looks good, the solution should at least work until someone starts to allocate tens of thousands of ports after another. But currently I think we do nothing of the sort, so hopefully it's fine.

Just one comment.

**Comment by jadolg** [comment]
> Thanks, looks good, the solution should at least work until someone starts to allocate tens of thousands of ports after another. But currently I think we do nothing of the sort, so hopefully it's fine.
> 
> Just one comment.

It's a compromise. We could use the mutex only for generating the port but there's no guarantee that the next won't be the same number if it's generated before the server has bound it. Better safe than sorrow.

**Review by juliusv — APPROVED** [review]
Thanks :+1:

**Pull request merged**